src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1377,28 +1377,58 @@ namespace Microsoft.FSharp.Collections
                       invalidArg "source" (SR.GetString(SR.notEnoughElements))
                   while e.MoveNext() do
                       yield e.Current }
-
+                           
         [<CompiledName("Last")>]
         let last (source : seq<_>) =
             checkNonNull "source" source
-            use e = source.GetEnumerator()
-            if e.MoveNext() then
-                let mutable res = e.Current
-                while (e.MoveNext()) do res <- e.Current
-                res
-            else
-                invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
-
+            match source with
+            | :? ('T[]) as a -> 
+                if a.Length = 0 then invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                else a.[a.Length - 1]
+            | :? ('T IList) as a -> //ResizeArray and other collections 
+                if a.Count = 0 then invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+                else a.[a.Count - 1]
+            | :? ('T list) as a ->
+                let rec listLast (list: 'T list) = // copied from List.last which is not available here (compilation order)
+                    match list with
+                    | [x] -> x
+                    | [] -> invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString  
+                    | _ :: tail -> listLast tail // inlined ? tail recursive?                                   
+                listLast a             
+            | _ -> 
+                use e = source.GetEnumerator()
+                if e.MoveNext() then
+                    let mutable res = e.Current
+                    while (e.MoveNext()) do res <- e.Current
+                    res
+                else
+                    invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+        
         [<CompiledName("TryLast")>]
         let tryLast (source : seq<_>) =
             checkNonNull "source" source
-            use e = source.GetEnumerator()
-            if e.MoveNext() then
-                let mutable res = e.Current
-                while (e.MoveNext()) do res <- e.Current
-                Some res
-            else
-                None
+            match source with
```

Comment:
I think you could reasonably make one (internal) implementation that returns a ValueOption and then define each of `last` and `tryLast` in terms of this.  Then there is no real code duplication.

Reply:
I moved the implementations of` last` and `tryLast` for `List` and `Seq` to `Microsoft.FSharp.Primitives.Basics` and use `ValueOption`, see https://github.com/dotnet/fsharp/pull/7765/commits/3086c97f89f50c71a20a6abe3f907be7d02feb81